### PR TITLE
Remove rental dependency

### PR DESCRIPTION
**Pull request by Theodus**
The rental crate is no longer maintained, and it has been broken by recent releases of Rust.

### CODE REVIEW

**Review by That3Percent — CHANGES_REQUESTED** [review]
I think this may be _technically_ unsound. But the unsoundness gets into this territory: [https://without.boats/blog/vulnerabilities/](https://without.boats/blog/vulnerabilities/) where you almost need to go out of your way to break it.

The issue is that it may possible to specialize traits depending on whether a type is `'static` [https://aturon.github.io/blog/2017/07/08/lifetime-dispatch/](https://aturon.github.io/blog/2017/07/08/lifetime-dispatch/). That would make it _possible_, however unlikely, to write code with undefined behavior in safe Rust because `Document<'static>` is lying about it's lifetime.

I would rather we do something like this:
* store a `*const Document`
* write a property which loads from the ptr to return `&Document<'self>`
* In `Drop` do a proper load and drop.

Which I think removes any unsoundness issues. Adding @leoyvens for a second opinion as well.

**Comment by Theodus** [comment]
I've updated to use a `*const u8` type for the struct field, since we can't specify a lifetime that fits our needs.

**Comment by otaviopace** [comment]
Your branch is compiler and `master` isn't, which is nice 🎉 , but the tests seem to be failing, at least here locally.

I don't know if I forgot to set something up besides just doing `cargo test`.

```
failures:

---- graphql_utils::tests::obj_recursion stdout ----
thread 'graphql_utils::tests::obj_recursion' panicked at 'assertion failed: query.is_err()', node-plugin/lang/src/graphql_utils.rs:213:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- graphql_utils::tests::list_recursion stdout ----
thread 'graphql_utils::tests::list_recursion' panicked at 'assertion failed: query.is_err()', node-plugin/lang/src/graphql_utils.rs:220:9


failures:
    graphql_utils::tests::list_recursion
    graphql_utils::tests::obj_recursion

test result: FAILED. 72 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 12.17s

error: test failed, to rerun pass '-p cost-model --lib'
```

スクリーンショット 2022-01-20 13 07 20

File: node-plugin/lang/src/lib.rs [hunk 1]
```diff
@@ -21,25 +19,31 @@ use fraction::{BigFraction, GenericFraction, Sign};
...
 }
 
-pub struct CostModel {
-    data: rentals::CostModelData,
+unsafe impl Send for CostModel {}
+unsafe impl Sync for CostModel {}
+
+impl Drop for CostModel {
+    fn drop(&mut self) {
+        let _ = unsafe { Box::<Document>::from_raw(transmute(self.document())) };
```
Comment (otaviopace) [inline node-plugin/lang/src/lib.rs:45]: I don't know if it's unsafer to do this, but there's no need to use `transmute` here, this works as well:

```rust
Box::from_raw(self.document as *mut u8)
```

File: node-plugin/lang/src/lib.rs [hunk 2]
```diff
@@ -21,25 +19,31 @@ use fraction::{BigFraction, GenericFraction, Sign};
...
-            text: String,
-            document: Document<'text>,
-        }
-    }
+pub struct CostModel {
+    // Rust does not have a memory model, nor does it have a proper `uintptr_t` equivalent. So a
+    // `*const u8` is used here, since the C99 standard explicitly allows casting from a C `char`
+    // the type of an object being accessed through a pointer (C99 §6.5/7). This assumes that all
+    // platforms being targeted will define C `char` as either a signed or unsigned byte.
+    document: *const u8,
```
Comment (leoyvens) [inline node-plugin/lang/src/lib.rs:33]: Why not `*const Document`?

**Comment by Theodus** [comment]
> Your branch is compiling and `master` isn't, which is nice tada , but the tests seem to be failing, at least here locally.

These tests have been failing since I joined E&N

File: node-plugin/lang/src/lib.rs [hunk 3]
```diff
@@ -130,21 +134,23 @@ impl fmt::Display for CompileError {
...
-            substitute_globals(&mut doc, &globals).map_err(|_| CompileError::Unknown)?;
-
-            Ok(doc)
-        })
-        .map_err(|e: rental::RentalError<CompileError, _>| e.0)?;
-
-        Ok(CostModel { data })
+        let globals = parse_vars(globals).map_err(CompileError::GlobalsParseError)?;
+        substitute_globals(&mut document, &globals).map_err(|_| CompileError::Unknown)?;
+        let document = unsafe { transmute(Box::into_raw(Box::new(document))) };
```
[the remark this answers is not in the document]
Comment (Theodus) [reply node-plugin/lang/src/lib.rs]: Thanks for pointing this out. I didn't realize that the Rust `as` keyword allows doing this without transmute. Fixed.

File: node-plugin/lang/src/lib.rs (hunk 1 above)
Comment (Theodus) [reply node-plugin/lang/src/lib.rs]: Using `as` seems to be an overall better alternative to the `transmute` sledgehammer.

**Review by otaviopace — APPROVED** [review]

**Pull request merged**